src/transformers/models/atlas/modeling_atlas.py

```diff
@@ -0,0 +1,380 @@
+        self.retriever = retriever
+        self.generator = generator
+
+    # todo
+    # - make it possible to separate the retrieval and generation steps
+    def forward(
+        self,
+        input_ids,
+        attention_mask,
+        labels,
+        query_input_ids,
+        query_attention_mask,
+        decoder_input_ids=None,
+        top_k=5,
+
+        # input_ids=None,
+        # attention_mask=None,
+        # encoder_hidden_states=None,
+        # encoder_attention_mask=None,
+        # inputs_embeds=None,
+        # head_mask=None,
+        # cross_attn_head_mask=None,
+        # past_key_values=None,
+        # use_cache=None,
+        # output_attentions=None,
+        # output_hidden_states=None,
+        # return_dict=None,
+    ):
+        bsz = len(input_ids)
+
+        query_hidden_states = self.retriever(input_ids=query_input_ids, attention_mask=query_attention_mask)
+        query_hidden_states_numpy = query_hidden_states.cpu().detach().numpy()
+
+
+        generator_tokens, retriever_tokens = self.retriever_index(query_hidden_states_numpy, input_ids, top_k)
+        generator_input_ids = generator_tokens["input_ids"]
+        generator_attention_mask = generator_tokens["attention_mask"].bool()
+
+        n_context_training = min(top_k, generator_input_ids.size(1))
+        cfg = self.generator.encoder.config
+        cfg.bsz = generator_input_ids.size(0)
+        cfg.n_context = n_context_training
+    
+        generator_input_ids_training = generator_input_ids[:, :n_context_training].contiguous()
+        generator_attention_mask_training = generator_attention_mask[:, :n_context_training].contiguous()
+
+        generator_input_ids_training = generator_input_ids_training.view(generator_input_ids.size(0), -1)
+        generator_attention_mask_training = generator_attention_mask_training.view(generator_attention_mask.size(0), -1)
+
+        generator_output = self.generator(
+            input_ids=generator_input_ids_training,
+            attention_mask=generator_attention_mask_training,
+            decoder_input_ids=decoder_input_ids,
+            labels=labels,
+            use_cache=False,
+        )
+
+        reader_loss = generator_output[0]
+
+        train_retriever = self.config.query_side_retriever_training and self.training
+
+        retriever_loss = None
+        if train_retriever:
+
+            query_emb = self.retriever(input_ids=query_input_ids, attention_mask=query_attention_mask, is_passages=False)
+            retriever_tokens = {k: v.reshape(-1, v.size(-1)) for k, v in retriever_tokens.items()}
+
+            passage_emb = self.retriever(**retriever_tokens, is_passages=True).to(query_emb)
+            passage_emb = passage_emb.view(bsz, -1, passage_emb.size(-1))
+            retriever_score = torch.einsum("id, ijd->ij", [query_emb, passage_emb])
+
+            if decoder_input_ids is None:
+                decoder_input_ids = self.generator._shift_right(labels)
```

Comment:
Labels are usually the final answers expected from the model. In the [Atlas implementation](https://github.com/facebookresearch/atlas/blob/4b6df19687210e4c8e9df2795664195041f9e45c/src/atlas.py#L229) the decoder_input_ids inlcude the query also.